nuxeo-compound-documents/src/main/java/org/nuxeo/compound/documents/CompoundDocumentServiceImpl.java

```diff
@@ -0,0 +1,98 @@
+ * limitations under the License.
+ *
+ */
+package org.nuxeo.compound.documents;
+
+import static org.nuxeo.compound.documents.CompoundDocumentConstants.COMPOUND_DOCTYPE;
+import static org.nuxeo.compound.documents.CompoundDocumentConstants.COMPOUND_FOLDER_DOCTYPE;
+import static org.nuxeo.runtime.model.Descriptor.UNIQUE_DESCRIPTOR_ID;
+
+import java.io.IOException;
+import java.util.Comparator;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipFile;
+
+import org.apache.commons.io.FilenameUtils;
+import org.nuxeo.common.utils.Path;
+import org.nuxeo.ecm.core.api.Blob;
+import org.nuxeo.ecm.core.api.CoreSession;
+import org.nuxeo.ecm.core.api.DocumentModel;
+import org.nuxeo.ecm.core.api.NuxeoException;
+import org.nuxeo.ecm.core.api.impl.blob.ZipEntryBlob;
+import org.nuxeo.ecm.platform.filemanager.api.FileImporterContext;
+import org.nuxeo.ecm.platform.filemanager.api.FileManager;
+import org.nuxeo.runtime.api.Framework;
+import org.nuxeo.runtime.model.DefaultComponent;
+
+/** @since 2021.0 */
+public class CompoundDocumentServiceImpl extends DefaultComponent implements CompoundDocumentService {
+
+    protected static final String XP_IGNORED_FILES = "ignoredFiles";
+
+    @Override
+    public DocumentModel createCompoundDocument(CoreSession session, String parent, Blob archive) {
+        String compoundDocName = FilenameUtils.removeExtension(archive.getFilename());
+        DocumentModel compoundDoc = session.createDocumentModel(parent, compoundDocName, COMPOUND_DOCTYPE);
+        compoundDoc = session.createDocument(compoundDoc);
+        createCompoundDocument(compoundDoc, archive);
+        return compoundDoc;
+    }
+
+    protected void createCompoundDocument(DocumentModel compoundDoc, Blob archive) {
+        try (ZipFile zip = new ZipFile(archive.getFile())) {
+            zip.stream()
+               .filter(this::isAllowedEntry)
+               .sorted(Comparator.comparing(ZipEntry::getName))
+               .forEach(entry -> createEntry(compoundDoc, archive.getFilename(), zip, entry));
+        } catch (IOException e) {
+            String message = String.format(
+                    "Failed to create CompoundDocument: %s. something went wrong with archive: %s.",
+                    compoundDoc.getName(), archive.getFilename());
+            throw new NuxeoException(message, e);
+        }
+    }
+
+    protected boolean isAllowedEntry(ZipEntry entry) {
+        Path path = new Path(entry.getName());
+        IgnoredFilesDescriptor ignoredFiles = getDescriptor(XP_IGNORED_FILES, UNIQUE_DESCRIPTOR_ID);
+        return ignoredFiles.isNotMatching(path);
+    }
+
+    protected void createEntry(DocumentModel compoundDoc, String archiveName, ZipFile zip, ZipEntry entry) {
+        Path entryPath = new Path(entry.getName());
+        Path entryName = entryPath.removeLastSegments(1);
+        String parentPath = compoundDoc.getPath().append(entryName).toString();
+        String docName = entryPath.lastSegment();
```

Comment:
This vars are hardly understandable, `entryName` seems to be a relative path, `docName` is the actual `entryName`, ...

What about:
```suggestion
        Path entryPath = new Path(entry.getName());
        Path entryParentPath = entryPath.removeLastSegments(1);
        String parentPath = compoundDoc.getPath().append(entryParentPath).toString();
        String entryName = entryPath.lastSegment();
```
Would it be more accurate?
(@kevinleturc )

Reply:
Yes indeed. But it is a bit strange to have two kinds of Path, may the following be better?
```suggestion
        Path entryPath = new Path(entry.getName());
        Path entryParentPath = entryPath.removeLastSegments(1);
        String parentDocPath = compoundDoc.getPath().append(entryParentPath).toString();
        String entryDocName = entryPath.lastSegment();
```